exp/controllers/machinepool_controller_noderef.go

```diff
@@ -173,53 +173,58 @@ func (r *MachinePoolReconciler) deleteRetiredNodes(ctx context.Context, c client
 	return nil
 }
 
-func (r *MachinePoolReconciler) getNodeReferences(ctx context.Context, c client.Client, providerIDList []string) (getNodeReferencesResult, error) {
+func (r *MachinePoolReconciler) getNodeReferences(ctx context.Context, c client.Client, providerIDList []string, owner string) (getNodeReferencesResult, error) {
 	log := ctrl.LoggerFrom(ctx, "providerIDList", len(providerIDList))
 
+	providerIDMap := map[string]bool{}
+	for _, providerID := range providerIDList {
+		pid, err := noderefutil.NewProviderID(providerID)
+		if err != nil {
+			log.V(2).Info("Failed to parse ProviderID, skipping", "err", err, "providerID", providerID)
+			continue
+		}
+		providerIDMap[pid.ID()] = true
+	}
+
 	var ready, available int
-	nodeRefsMap := make(map[string]corev1.Node)
+	var nodeRefs []corev1.ObjectReference
 	nodeList := corev1.NodeList{}
 	for {
 		if err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue)); err != nil {
 			return getNodeReferencesResult{}, errors.Wrapf(err, "failed to List nodes")
 		}
 
-		for _, node := range nodeList.Items {
+		for idx := range nodeList.Items {
+			node := nodeList.Items[idx]
+			if node.ObjectMeta.Annotations[clusterv1.OwnerKindAnnotation] != "MachinePool" || node.ObjectMeta.Annotations[clusterv1.OwnerNameAnnotation] != owner {
```

Comment:
[MachinePool](https://github.com/kubernetes-sigs/cluster-api/blob/c6380c53e27a927a7cf22fd43b43c357a57925a7/exp/controllers/machinepool_controller_noderef.go#L117) annotations for the nodes are set after we fetch the nodeRefs.

Reply:
hmm…might need to wait for the full provider ID fix in that case…

Reply:
Yeah i was thinking the same. In the current code why are we not using the full provider id by the way ?

Reply:
I don't think there's a good reason, see https://github.com/kubernetes-sigs/cluster-api/pull/4913#issuecomment-878041148

I think we just need to do the work to fix/normalize that everywhere